src/lib/device-online-state/index.ts

```diff
@@ -0,0 +1,280 @@
+		this.rsmq
+			.createQueue({ qname: DeviceOnlineStateManager.EXPIRED_QUEUE })
+			.catch((err: Error) => {
+				if (err.name !== 'queueExists') {
+					throw err;
+				}
+			})
+			.then(() => {
+				this.consume();
+			});
+	}
+
+	// event emitters
+	private deviceOnline(uuid: string) {
+		this.emit('change', uuid, DeviceOnlineStates.Online);
+	}
+
+	private deviceOffline(uuid: string) {
+		this.emit('change', uuid, DeviceOnlineStates.Offline);
+	}
+
+	private deviceTimeout(uuid: string) {
+		this.emit('change', uuid, DeviceOnlineStates.Timeout);
+	}
+
+	private consume() {
+		let delayBeforeNextConsume = 1;
+
+		// pull a message from the queue...
+		this.rsmq
+			.receiveMessage({
+				qname: DeviceOnlineStateManager.EXPIRED_QUEUE,
+				vt: 30, // prevent other consumers seeing the same message (if any) for 30 secs (prevents multiple API agents from processing it)...
+			})
+			.tap((msg: any) => {
+				// no message! this is thrown to restart the consume process...
+				if (!msg || !msg.id) {
+					delayBeforeNextConsume = 1000;
+					throw new NoMessagesToConsumeSignal();
+				}
+			})
+			.then((msg: any) => {
+				const { id, message } = msg;
+
+				return Bluebird.try(() => {
+					const { uuid, nextState } = JSON.parse(message);
+
+					// raise and event for the state change...
+					switch (nextState) {
+						case DeviceOnlineStates.Timeout:
+							this.scheduleChangeOfStateForDevice(
+								uuid,
+								DeviceOnlineStates.Timeout,
+								DeviceOnlineStates.Offline,
+								15, // put the device into a timeout state for 15 sec if it misses it's scheduled heartbeat window... then mark as offline
+							);
+							this.deviceTimeout(uuid);
+							break;
+						case DeviceOnlineStates.Offline:
+							this.deviceOffline(uuid);
+							break;
+						default:
+							// nothing should be raised here...
+							break;
+					}
+				}).then(() =>
+					this.rsmq.deleteMessage({
+						qname: DeviceOnlineStateManager.EXPIRED_QUEUE,
+						id: id,
+					}),
+				);
+			})
+			.catch(NoMessagesToConsumeSignal, noop)
+			.catch((err: Error) => {
+				captureException(
+					err,
+					'DeviceStateManager ==> Error when consuming queue',
+				);
+			})
+			.then(() => {
+				// schedule the next consume...
+				setTimeout(() => {
+					this.consume();
+				}, delayBeforeNextConsume);
+			});
+	}
+
+	private scheduleChangeOfStateForDevice(
+		uuid: string,
+		currentState: DeviceOnlineStates,
+		nextState: DeviceOnlineStates,
+		inSeconds: number,
+	) {
+		// remove the old queued state...
+		return this.redis
+			.getAsync(`${DeviceOnlineStateManager.REDIS_NAMESPACE}:${uuid}`)
+			.then(value => {
+				const { id } = JSON.parse(value) as { id: string };
+
+				if (id) {
+					return this.rsmq
+						.deleteMessage({
+							qname: DeviceOnlineStateManager.EXPIRED_QUEUE,
+							id,
+						})
+						.return();
+				}
+			})
+			.catch(() => noop) // ignore errors when deleting the old queued state and queue the new one...
```

Comment:
Do you really mean to return a noop function here?

Reply:
Yes. If there is an error clearing the existing value, then it just needs to carry on and set the new one; it isn't an error that we need to worry about.

Reply:
I think @Page- means this should be `.catch(noop)` instead, because currently as it is it will return the `noop` function.